runtime/shared_common/SCImplementedAPI.cpp

```diff
@@ -490,9 +501,24 @@ j9shr_classStoreTransaction_start(void * tobj, J9VMThread* currentThread, J9Clas
 			/* For class loaded from modules that entryIndex is -1. classPathEntries can be NULL. */
 			void* cpExtraInfo = NULL;
 			UDATA infoFound = 0;
+			J9ClassPathEntry* firstcpe = NULL;
 
 			if (NULL != classPathEntries) {
-				cpExtraInfo= classPathEntries->extraInfo;
+				/*
+				 * class path entries are never freed during j9shr_classStoreTransaction.
+				 * For bootstrap class loader class path entries are never freed.
+				 * For non-bootstrap class loader, its classPathEntries is only used by the shared class code
+				 * (j9shr_classStoreTransaction, hookFindSharedClass and shared.c). They are all protected by
+				 * SharedClassURLClasspathHelperImpl.urlcpReadWriteLock.
+				 *
+				 * It is safe to cache the first entry pointer.
+				 */
+				if (useLoaderCpEntries) {
+					firstcpe = classloader->classPathEntries[0];
```

Comment:
I am avoid taking the cpEntriesMutex in this function. All other places using bootstrap classloader->classPathEntries is only to check if it is NULL. they won't read the actual class path entry. I am using classloader->classPathEntries[0] here to get the latest value of classloader->classPathEntries in case it is re-allocated.  I assume this line (reading one address) is atomic.

Reply:
I don't think it's atomic. This thread could get the old value classloader->classPathEntries, the other thread could then free it, and then this thread reads the first entry from freed memory.